toolchain/lex/tokenized_buffer.cpp

```diff
@@ -352,32 +376,92 @@ class [[clang::internal_linkage]] TokenizedBuffer::Lexer {
     CARBON_DCHECK(source_text.substr(position).startswith("//"));
 
     // Any comment must be the only non-whitespace on the line.
-    if (set_indent_) {
+    const auto* line_info = current_line_info();
+    if (LLVM_UNLIKELY(position != line_info->start + line_info->indent)) {
       CARBON_DIAGNOSTIC(TrailingComment, Error,
                         "Trailing comments are not permitted.");
 
       emitter_.Emit(source_text.begin() + position, TrailingComment);
     }
 
     // The introducer '//' must be followed by whitespace or EOF.
-    if (position + 2 < static_cast<ssize_t>(source_text.size()) &&
+    if (LLVM_UNLIKELY(position + 2 <
+                      static_cast<ssize_t>(source_text.size())) &&
         !IsSpace(source_text[position + 2])) {
       CARBON_DIAGNOSTIC(NoWhitespaceAfterCommentIntroducer, Error,
                         "Whitespace is required after '//'.");
       emitter_.Emit(source_text.begin() + position + 2,
                     NoWhitespaceAfterCommentIntroducer);
     }
 
-    // Use the current line info to jump to the end of the line.
-    position = current_line_info()->start + current_line_info()->length;
-    // This may be the end of the file in which case we immediately return.
-    if (position == static_cast<ssize_t>(source_text.size())) {
-      // Finished lexing.
-      return;
+    // Skip over this line.
+    ssize_t line_index = line_index_;
+    ++line_index;
+    position = buffer_.line_infos_[line_index].start;
+
+    // A very common pattern is a long block of comment lines all with the same
+    // indent and comment start. We skip these comment blocks in bulk both for
+    // speed and to reduce redundant diagnostics if each line has the same
+    // erroneous comment start like `//!`.
+    //
+    // When we have SIMD support this is even more important for speed, as short
+    // indents can be scanned extremely quickly with SIMD and we expect these to
+    // be the dominant cases.
+    //
+    // TODO: We should extend this to 32-byte SIMD on platforms with support.
+    constexpr int MaxIndent = 13;
+    const int indent = line_info->indent;
+    const ssize_t first_line_start = line_info->start;
+    auto skip_to_next_line = [this, indent, &line_index, &position] {
+      // We're guaranteed to have a line here even on a comment on the last line
+      // as we ensure there is an empty line structure at the end of every file.
+      ++line_index;
+      auto* next_line_info = &buffer_.line_infos_[line_index];
+      next_line_info->indent = indent;
+      position = next_line_info->start;
+    };
+    if (CARBON_USE_SIMD &&
+        position + 16 < static_cast<ssize_t>(source_text.size()) &&
+        indent <= MaxIndent) {
+#if __x86_64__
+      // Load a mask based on the amount of text we want to compare.
+      auto mask = prefix_masks[indent + 3];
```

Comment:
I think this should be `+ 2` if the character after `//` was not a space. As-is, I think we would get one diagnostic for
```carbon
//hello
//hello
```
but two for
```carbon
//hello
//world
```

Reply:
Enhanced test and fixed.